activerecord/lib/active_record/destroy_later.rb

```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  # Automatically destroy records after a specified amount of time.
+  #
+  #   # Destroy all exports 30 days after creation
+  #   class Export < ApplicationRecord
+  #     destroy_later after: 30.days
+  #   end
+  #
+  #   # Destroy exports 30 days after completing them
+  #   class Export < ApplicationRecord
+  #     destroy_later after: 30.days, if: -> { completed_previously_changed? && completed? }
+  #   end
+  #
+  # Optionally pass the name of an instance method in +ensuring:+. If the given method returns false at the
+  # scheduled destroy time, the record will not be destroyed.
+  #
+  #   # Destroy exports 30 days after completing them, ensuring they're still marked as completed at destroy time
+  #   class Export < ApplicationRecord
+  #     destroy_later after: 30.days, if: -> { completed_previously_changed? && completed? }, ensuring: :completed?
+  #   end
+  module DestroyLater
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      # :call-seq: destroy_later(after:, if: nil, ensuring: nil)
+      def destroy_later(after:, ensuring: nil, **options)
```

Comment:
I'm not sure how we (or even I) feel about one vs the other, but FWIW I believe the officially-recommended way to do this is:

```ruby
def destroy_later(after:, if: nil, ensuring: nil)
  if_ = binding.local_variable_get(:if)
```

Reply:
I like that this results in an `ArgumentError` if you pass an unsupported option. I went without the extra local variable.